lib/thrift/generator/client.ex

```diff
@@ -0,0 +1,270 @@
+
+                    case response do
+                      [exception] ->
+                        {:error, {:exception, exception}}
+
+                      [] ->
+                        # This case is when we have a void return on the
+                        # remote RPC
+                        {:ok, nil}
+                    end
+
+                  {%{success: success}, ""} when not is_nil(success) ->
+                    {:ok, success}
+
+                  {resp, extra} ->
+                    {:error, {:extraneous_data, resp, extra}}
+                end
+
+              {:ok, {^sequence_id, :exception, ^rpc_name, response}} ->
+                exc = read_t_application_exception(response, %TApplicationException{})
+                {:error, {:exception, exc}}
+
+              {:ok, {^sequence_id, _, mismatched_rpc_name, _}} ->
+                message = "The server replied to #{mismatched_rpc_name}, but we sent #{rpc_name}"
+                exc = %TApplicationException{
+                  message: message,
+                  type: BinaryBase.exception_type(3)}
+                {:error, {:exception, exc}}
+
+              {:ok, {mismatched_sequence_id, _, ^rpc_name, _}} ->
+                message = "Invalid sequence id. The client sent #{sequence_id}, but the server replied with #{mismatched_sequence_id}"
+
+                exc = %TApplicationException{message: message,
+                                             type: BinaryBase.exception_type(4)
+                                            }
+                {:error, {:exception, exc}}
+
+              {:ok, {_mismatched_sequence_id, _mismatched_rpc_name, _}} ->
+                message = "Both sequence id and rpc name are wrong. The server is extremely uncompliant."
+                exc = %TApplicationException{message: message,
+                                             type: :sequence_id_and_rpc_name_mismatched}
+                {:error, {:exception, exc}}
+
+              {:error, _}=err ->
+                err
+            end
+          end
+
+          defp read_t_application_exception(
+                <<11::size(8),
+                1::16-unsigned,
+                message_size::32-signed,
+                message::binary-size(message_size),
+                rest::binary>>, accum) do
+            # read the message string
+            read_t_application_exception(rest, Map.put(accum, :message, message))
+          end
+          defp read_t_application_exception(
+                <<8::size(8),
+                2::16-unsigned,
+                type::32-signed,
+                rest::binary>>, accum) do
+            # read the type
+            exception_type = BinaryBase.exception_type(type)
+            read_t_application_exception(rest, Map.put(accum, :type, exception_type))
+          end
+          defp read_t_application_exception(<<0>>, accum) do
+            # read the field stop and return
+            accum
+          end
```

Comment:
This parser should return {value, rest} or :error for consistency, and because where it is called up above, we should handle the case where the exception can not be decoded properly and return a protocol error (type 7).